nncf/common/pruning/utils.py

```diff
@@ -395,6 +392,99 @@ def get_operator_metatype_by_op_name(self, op_name: str):
         return None
 
 
+class PruningAnalysisReason(Enum):
+    """
+    Enum of possible pruning analysis decisions reasons.
+    """
+
+    IGNORED_SCOPE = 'node in ignored scope'
+    FIRST_CONV = 'this scope is one of the first convolutions'
+    LAST_CONV = 'this scope is convolution with output which directly affects model output dimensions'
+    GROUP_CONV = 'this scope is grouped convolution'
+    DOWNSAMPLE_CONV = 'this scope is convolution with downsample'
+    MODEL_ANALYSIS = 'of model analysis'
+    DIMENSION_MISMATCH = 'of dimension mismatch'
+    CLOSING_CONV_MISSING = 'closing convolution missing'
+    IN_GROUP_OF_UNPRUNABLE = 'is in the group with non prunable layers'
+
+    @classmethod
+    def message(cls, node_name: str, decision: Optional['PruningAnalysisDecision']) -> str:
+        """
+        Returns the node pruning analysis decisions in a human-readable format.
+
+        :param node_name: Name of given node.
+        :param decision: Pruning analysis decision for given node.
+        :return: Pruning analysis decision in a human-readable format.
+        """
+        prefix = f'Ignored adding Weight Pruner in: {node_name}'
+        reasons = decision.reasons
+        if not reasons:
+            return prefix
+        # Filter messages
+        if cls.MODEL_ANALYSIS in reasons and cls.CLOSING_CONV_MISSING in reasons:
```

Comment:
After `analyse_model_before_pruning`, which is a first step, all nodes in graph have `MODEL_ANALYSIS` as a reason + `join` doesn't remove any of the previous reasons. Do we even need this check?

Reply:
All nodes which removed by model analysis will be removed by symbolic mask prop, thus I decided to filter this

Reply:
Should be like

```suggestion
        if len(reasons) > 1 and cls.CLOSING_CONV_MISSING in reasons:
            reasons.remove(cls.CLOSING_CONV_MISSING)
```

Reply:
> All nodes which are removed by model analysis will be removed by symbolic mask prop

Sounds like we don't need a model analysis step. Is this true?

Reply:
It's not, because as description says:

"This check is added after model analyzer and per pruned layer check because mask propagation algo assumes every prunable layer children subtree accept prunable input (this checked by model analyzer) and all prunable layers are supported by current pruning algo implementation (this checked by per pruned layer check) thus all "closing" convolution is reachable or, in another words, no cases when symbolic mask is stopped in the way from producer to "closing" convolution. (In case of general input, it can be two branches from a producer, one prunable and one not, and symbolic mask prop algo just skipped non prunable branch and label producer as "prunable", which is wrong)."